XUtils.c

```diff
@@ -374,6 +374,20 @@ double sumPositiveValues(const double* array, size_t count) {
    return sum;
 }
 
+size_t countDigit(size_t n, size_t base) {
+   assert(base > 1);
+   if (n == 0) {
+      return 1;
+   }
+
+   size_t res = 0;
+   while (n != 0) {
+      n /= base;
```

Comment:
Looks like a naive implementation, but using division is slow. I don't know what kind of performance you and @BenBE are aiming for.

Reply:
The version in the link I pointed too gives a constant time version.

Given most numbers here should be fairly small, the number of iterations here is kinda small too. Thus unless division is a lot slower than this might still be faster compared to the lookup variant.

Some hard numbers here might be good, maybe when comparing even compare with the float variant as a reference.

But, given how little this function is called (once/twice per process maybe) I don't think it's worth optimizing for one or two CPU cycles here.

Reply:
@BenBE No, I mean, it's better to use multiplication even when you are optimizing this function for small numbers. You are not converting base anyway and thus you don't need remainders for every division.

I'm here to contribute my version: It generates slightly smaller machine code (for x86). 🙂 

https://godbolt.org/z/6aWvcc3z8

```c
size_t countDigit_2(size_t n, size_t base) {
   assert(base > 1);

   size_t res = 1;
   size_t limit = base;
   while (n >= limit) {
      res++;
      if (limit > SIZE_MAX / base) {
         break;
      }
      limit *= base;
   }
   return res;
}
```

Reply:
@Explorer09 Good catch/suggestion. Let's take it a another (favoured) candidate.

Also, your's can be shortened somewhat:

```c
size_t countDigit_2(size_t n, size_t base) {
   assert(base > 1);

   size_t res = 1;
   for (size_t limit = base; n >= limit; limit *= base) {
      res++;
      if (limit > SIZE_MAX / base) {
         break;
      }
   }
   return res;
}
```

The `SIZE_MAX / base` is a runtime constant, that the compiler will likely calculate once before the loop (if `-fmove-loop-invariants` is used, which is the default for `-O1` and above, but not `-Og`). This could be made explicit:

```c
size_t countDigit_2(size_t n, size_t base) {
   assert(base > 1);

   size_t res = 1;
   for (size_t limit = base, max_limit = SIZE_MAX / base; n >= limit; limit *= base) {
      res++;
      if (limit > max_limit) {
         break;
      }
   }
   return res;
}
```
In this case, even `-O0` and `-Og` would benefit from avoiding this additional division.

With `-fipa-cp` the division is moved to compile time, which is the default for `-O2`, `-O3`, and `-Os`.

Reply:
Ah, wasn't aware of that builtin. Thx.

Reply:
Just to mention I found a missed optimization when working with this `countDigits` function:
[GCC bug report](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=117529), [LLVM bug report](https://github.com/llvm/llvm-project/issues/115683)